### ISSUE TITLE
Proposal: Break down `Router` into components 

### ISSUE THREAD

**Issue by Altair-Bueno** [tokio-rs/axum#1318]
### Motivation

Current v0.6.0 introduces some unfortunate breaking changes

- Nested routers no longer delegates to the outer router's fallback
- State must be duplicated for nested routers

### Proposed solution

Break down `Router` into `Router`, `State` and `Fallback` using Tower's `Layer` and `Service` traits

```rs
// Request -> AxumResponse
let service = ServiceBuilder::new()
  // Transforms the raw request into something that `Router` and `Fallback` understand.
  // Request -> Inner::call(AxumRequest<RequestParts, State>
  .layer(State::default())
  // Transforms `Result<Response, NotFound>` (returned by `Router`) into 
  // `Result<Response, Infalible>`
  // AxumRequest<RequestParts, State> -> Inner::(AxumRequest { RequestParts, State }).unwrap_or(DefaultFallback)
  .layer(Fallback::default())
  // The actual routing and handlers live as a separate service, whose input/output is 
  // adapted by other layers
  // AxumRequest<RequestParts, State> -> Result<AxumResponse, NotFound>
  .service(Router::new().route("/", get(foo)));

axum::Server::bind("127.0.0.1:8080".parse()?)
  .serve(service).await?;
``` 

This creates a separation of concerns, meaning that we can nest routers without worrying about state or fallbacks

### Pros

- Because it's based on Tower, it's possible to interchange layers or add more middleware in between
- Probably helps with https://github.com/tokio-rs/axum/pull/1054

### Cons

- Creating the main router is more verbose. We can fix this by creating an `AxumApp` router that hides everything under a builder pattern

```rs
// Uses default State (()) and default fallback
AxumApp::new().router(router).build()
```

### Additional notes

I don't know if there is any technical limitation that makes this composition pattern impossible. I've spent the last couple of days looking at the `Router` source code but it's a bit hard to follow along (i have little experience with macros). Given that v0.6.0 is around the corner, i think is more productive to publish this issue now

**Comment by davidpdrsn** [comment]
> Nested routers no longer delegates to the outer router's fallback

It is not clear whether thats good or bad. I've heard from people who don't like it and people who prefer it. Personally I don't mind it since it fixes many inconsistencies and avoids some difficult questions.

For example if I write

```rust
let api = Router::new().layer(some_middleware);
let app = Router::new().nest("/api", api).layer(some_other_middleware);
```

and then call `/api/doesnt/exist` does that go through `some_middleware`? What about `some_other_middleware`? What happens if the fallback is set on `api` but not `app`? Which middleware then get run?

By saying that `nest` swallows all matching requests we avoid this problem. `some_middleware` will be called because from the perspective of `api` its just a normal request.

> Break down Router into Router, State and Fallback using Tower's Layer and Service traits

This feels like a large change that could complicate things a lot. The value in making that split isn't clear to me. Can you elaborate?

"separation of concerns" means different things to different people and has no value in and off it self. What is it that you cannot do today, that this change enables?

I think part of axum's appeal is that conceptually its design is simple and familiar to people (a la [express](https://expressjs.com) or [sinatra](https://sinatrarb.com)). While I don't fully understand your proposal I worry it makes things more foreign to people. Keep in mind that many axum users are people who are new to Rust. Hiding stuff underneath builders is not a viable option. Those are often leaky abstractions that don't actually make things simpler, in my experience.

**Comment by Altair-Bueno** [comment]
> It is not clear whether thats good or bad. 

This change would allow both behaviours. The router just routes to handlers and middleware. It doesn't know anything about state nor Fallbacks. That's defined on another layer

- If no `Fallback` layer is provided, the `NotFound` error would be returned to the parent router
- If a `Fallback` layer is provided, the `NotFound` error would be handled before reaching the parent router

Following your example, it would depend on how you set up your layers

```rs
// /api/doesnt/exist
// some_middleware && some_other_middleware
let service = ServiceBuilder::new()
  .layer(State::default())
  .layer(some_middleware)
  .layer(some_other_middleware)
  .layer(Fallback::default())
  .service(
	Router::new()
      // Note that this router neither defines its own State nor Fallback
      .nest("/api",Router::new())
  );

// some_middleware && !some_other_middleware
let service = ServiceBuilder::new()
  .layer(State::default())
  .layer(some_middleware)
  .layer(Fallback::default())
  .service(
	Router::new()
      .nest("/api",Router::new().layer(some_other_middleware))
  );
```

> What is it that you cannot do today, that this change enables?

- The option to fall back to a parent `Fallback`
- State passes through to nested routers. 
- Due to Tower's nature, the ability to hook middleware in between `State` and `Router` (`Fallback` would be a `map_err` middleware)

> While I don't fully understand your proposal I worry it makes things more foreign to people

I would say this proposal isn't mature enough right now, but after talking with you last night I thought it may be good enough to at least open an issue with the proposal

As you said, using `ServiceBuilder` is not probably not ideal for new users. Nontheless, `AxumApp` doesn't feel alien to me. Flask already does this:

- `Flask` is used to create the actual root application. It's a router with a couple of add- ons [^1]
- `Blueprint` is used for nested routes. It's just the router layer [^2]

Note: As always, I'm not good at choosing names

[^1]: https://flask.palletsprojects.com/en/2.2.x/quickstart/#a-minimal-application
[^2]: https://flask.palletsprojects.com/en/2.2.x/tutorial/views/

**Comment by davidpdrsn** [comment]
Sorry but I still don't see the value in this. Not even sure I follow how it would actually work.

Maybe you need to implement something and open a new proposal when you have something more concrete?

**Comment by davidpdrsn** [comment]
I think I'll close this for now for the reasons posted above.

We do wanna look into making state/fallback sharing more flexible and easier to understand. So feel free to share your opinions. See https://github.com/tokio-rs/axum/issues/1326.
